### PR TITLE
[Tailcall] Enable tailcall of virtual functions on x86 and ARM64.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

File: mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -7024,14 +7024,16 @@ is_supported_tail_call (MonoCompile *cfg, MonoMethod *method, MonoMethod *cmetho
...
+	gboolean const llvm_only = cfg->llvm_only;
 
 	g_assertf (call_opcode == CEE_CALL || call_opcode == CEE_CALLVIRT || call_opcode == CEE_CALLI, "%s (%d)", mono_opcode_name (call_opcode), (int)call_opcode);
 
 	if (       (fsig->hasthis && m_class_is_valuetype (cmethod->klass))  // This might point to the current method's stack. Emit range check?
 		|| (cmethod->flags & METHOD_ATTRIBUTE_PINVOKE_IMPL)
 		|| cfg->method->save_lmf
 		|| (cmethod->wrapper_type && cmethod->wrapper_type != MONO_WRAPPER_DYNAMIC_METHOD)
-		|| call_opcode == CEE_CALLI
+		|| (!llvm_only && (
```
Comment (jaykrell) [inline mono/mini/method-to-ir.c]: Is LLVM too willing? For example could we pass a stale local address as part of returning a struct by value? I'd think LLVM owns the ABI there and everywhere and so won't mess it up, so we probably really still have too many restrictions here.

**Comment by jaykrell** [comment]
Woohoo tests passed in one try (there was just one small followup upon the initial submit, probably just undoing more of the arm32 change but I don't remember)

**Comment by jaykrell** [comment]
A few relevant new tests broken, looking.

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
Ping?

File: mono/mini/mini-x86.c [hunk 2]
```diff
@@ -70,7 +70,42 @@ static mono_mutex_t mini_arch_mutex;
...
+	int bitset;
+	int push [callee_saved_regs_count];
+	int pop [callee_saved_regs_count];
+	int monolmf [callee_saved_regs_count + 1];          // +1 for EBP special case
+	int monolmf_offset [callee_saved_regs_count + 1];   // +1 for EBP special case
+} MonoX86CalleeSavedRegisters;
+
+const static MonoX86CalleeSavedRegisters
+callee_saved_regs = {
+	callee_saved_regs_count,                            // count
```
Comment (vargaz) [inline mono/mini/mini-x86.c]: What is the purpose of these changes ? Is the set of registers on x86 going to change in future ? Does it differ between OSes ? The new code looks harder to understand than the old one.

Comment (jaykrell) [reply mono/mini/mini-x86.c]: Factoring out the repetition makes it clear what is different vs. the same in each reptition. i.e. just the registers.

**Comment by jaykrell** [comment]
I'll close this and propose similar but cleaner and with some of the commits in separate PR.

**Comment by jaykrell** [comment]
In particular I want to save the arm changes in this branch for maybe another day.

**Pull request closed without merging**